### PR TITLE
DRA: test for structured parameters

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/pull/123516 got merged with some tests still missing due to time pressure. This PR adds those tests. One bug was found while doing so.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by pohly** [comment]
Unit testing depends on correct handling of `ResourceClaimParameters` and `ResourceClassParameters` by the fake client, which is currently broken because of a [wrong guess of the plural form](https://github.com/kubernetes/kubernetes/pull/123936). The [right solution](https://github.com/kubernetes/kubernetes/pull/119924) is still pending. Note sure yet how to handle this. Merge without the `dynamicresources_test.go` changes?

**Comment by pohly** [comment]
> The [right solution](https://github.com/kubernetes/kubernetes/pull/119924) is still pending. Note sure yet how to handle this. Merge without the `dynamicresources_test.go` changes?

I have [a workaround](https://github.com/kubernetes/kubernetes/pull/123936#issuecomment-1998082422). I'm currently waiting for https://github.com/kubernetes/kubernetes/pull/123932 to merge, then I will rebase.

**Comment by pohly** [comment]
Rebased, PR should be okay now.

/test pull-kubernetes-node-e2e-containerd-1-7-dra

[Looks](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123938/pull-kubernetes-node-e2e-containerd-1-7-dra/1768364746291023872) like a Flake.

/cc @bart0sh

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go [hunk 1]
```diff
@@ -152,6 +222,44 @@ var (
 		Obj()
 )
 
+func reserve(claim *resourcev1alpha2.ResourceClaim, pod *v1.Pod) *resourcev1alpha2.ResourceClaim {
+	return st.FromResourceClaim(claim).
+		ReservedFor(resourcev1alpha2.ResourceClaimConsumerReference{Resource: "pods", Name: pod.Name, UID: types.UID(pod.UID)}).
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:227]: Maybe you can have a `ReservedForPod` method?

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go [hunk 2]
```diff
@@ -256,6 +380,23 @@ func TestPlugin(t *testing.T) {
...
+							if claim.Name == claimName {
+								claim = claim.DeepCopy()
+								claim.Status.ReservedFor = inUseClaim.Status.ReservedFor
+							}
+							return claim
+						},
+					},
+				},
+			},
+		},
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:399]: TBH, I have zero idea what is happening here. There are way too many things hidden in abstractions.

what is podWithClaimName? why makes it structured? what does `change` receive?

https://abseil.io/resources/swe-book/html/ch12.html#:~:text=Tests%20and%20Code%20Sharing%3A%20DAMP%2C%20Not%20DRY

File: pkg/controller/resourceclaim/controller_test.go [hunk 3]
```diff
@@ -546,6 +615,14 @@ func allocateClaim(claim *resourcev1alpha2.ResourceClaim) *resourcev1alpha2.Reso
 	return claim
 }
 
+func structuredParameters(claim *resourcev1alpha2.ResourceClaim) *resourcev1alpha2.ResourceClaim {
+	claim = claim.DeepCopy()
+	// As far the controller is concerned, a claim was allocated by us if it has
+	// this finalizer. For testing we don't need to update the allocation result.
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/controller/resourceclaim/controller_test.go:621]: I'll add that in a separate PR for the API. I have another comment update pending.

File: pkg/controller/resourceclaim/controller_test.go [hunk 4]
```diff
@@ -293,6 +293,19 @@ func TestSyncHandler(t *testing.T) {
 			}(),
 			expectedMetrics: expectedMetrics{0, 0},
 		},
+		{
+			name:   "clear-reserved-delayed-allocation-structured",
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/controller/resourceclaim/controller_test.go:297]: Added (but not pushed yet - I'm not sure whether I'll be able finish today).

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 5]
```diff
@@ -1410,7 +1410,11 @@ func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleStat
 		}
 		state.informationsForClaim[index].allocation = allocation
 		state.informationsForClaim[index].allocationDriverName = driverName
+		// Strictly speaking, we don't need to store the full modified object.
+		// The allocation would be enough. The full object is useful for
+		// debugging and testing, so let's make it realistic.
 		claim = claim.DeepCopy()
+		claim.Finalizers = append(claim.Finalizers, resourcev1alpha2.Finalizer)
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1417]: All in memory. It's debatable whether it should store the whole "expected" claim at all. It's consistent, but not really required.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go (hunk 1 above)
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:227]: Added.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go (hunk 2 above)
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:399]: Pods are not different, only claims are (potentially different parameter reference, different status). The way how I arrived at some of these unit tests is that I copied existing ones, swapped out the claims for their counterparts and checked what the expected result should be. This gave me roughly coverage of the same scenarios that I had before.

I then thought about some more special cases that can only occur when using structured parameters.

I understand that it's hard to verify that this covers all corner cases. I checked code coverage, but that's only a partial answer.

Regarding "want changes": this is a helper to describe the expected changes more concisely. In this case, the only operation which has any visible effect in the things that are getting checked (cluster objects, assume cache, in-flight claims) is "prebind".

I think it is easier to see what the effect is when describing what changes. Otherwise the test case would have to list the full expected state for each operation and it would be very hard to see that prebind does something.

**Comment by pohly** [comment]
PR updated.

File: test/e2e/dra/deploy.go [hunk 6]
```diff
@@ -84,10 +88,56 @@ func NewNodes(f *framework.Framework, minNodes, maxNodes int) *Nodes {
 			nodes.NodeNames = append(nodes.NodeNames, node.Name)
 		}
 		framework.Logf("testing on nodes %v", nodes.NodeNames)
+
+		// Watch claims in the namespace. This is useful for monitoring a test
+		// and enables additional sanity checks.
+		claimInformer := resourceapiinformer.NewResourceClaimInformer(f.ClientSet, f.Namespace.Name, 100*time.Hour /* resync */, nil)
+		cancelCtx, cancel := context.WithCancelCause(context.Background())
```
Comment (alculquicondor) [inline test/e2e/dra/deploy.go:95]: use a ktesting context?

File: test/e2e/dra/dra.go [hunk 7]
```diff
@@ -128,8 +128,25 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 				// arbitrary types we can simply fake somthing here.
 				claim := b.externalClaim(resourcev1alpha2.AllocationModeWaitForFirstConsumer)
 				b.create(ctx, claim)
+
 				claim, err := f.ClientSet.ResourceV1alpha2().ResourceClaims(f.Namespace.Name).Get(ctx, claim.Name, metav1.GetOptions{})
 				framework.ExpectNoError(err, "get claim")
+
+				claim.Finalizers = append(claim.Finalizers, "e2e.test/delete-protection")
+				claim, err = f.ClientSet.ResourceV1alpha2().ResourceClaims(f.Namespace.Name).Update(ctx, claim, metav1.UpdateOptions{})
```
Comment (alculquicondor) [inline test/e2e/dra/dra.go:136]: should we retry Update calls? could there be conflicts?

File: test/e2e/dra/dra.go [hunk 8]
```diff
@@ -912,7 +929,7 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 							"NodeName":   gomega.Equal(nodeName),
 							"DriverName": gomega.Equal(driver.Name),
 							"ResourceModel": gomega.Equal(resourcev1alpha2.ResourceModel{NamedResources: &resourcev1alpha2.NamedResourcesResources{
-								Instances: []resourcev1alpha2.NamedResourcesInstance{{Name: "instance-0"}},
+								Instances: []resourcev1alpha2.NamedResourcesInstance{{Name: "instance-00"}},
```
Comment (alculquicondor) [inline test/e2e/dra/dra.go:932]: why this change?

**Review by alculquicondor — COMMENTED** [review]
kinda LGTM.
The current unit testing machinery is very hard to follow.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go [hunk 9]
```diff
@@ -535,6 +926,30 @@ func TestPlugin(t *testing.T) {
...
+						claim: func(in *resourcev1alpha2.ResourceClaim) *resourcev1alpha2.ResourceClaim {
+							out := in.DeepCopy()
+							out.Status.ReservedFor = []resourcev1alpha2.ResourceClaimConsumerReference{}
+							return out
+						},
+					},
+				},
+			},
+		},
 		"reserved-okay": {
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:953]: Gotcha. A test case was missing *and* the "if controller != nil" check above that was wrong (no controller for allocated claims, only `structuredParameters` is set), so it didn't work as intended.

I need to think a bit more about how to trigger this case in a real E2E scenario. There is a test case for classic DRA which is based on artificially creating the right timing conditions, but with structured parameters I can only think of a write error or some interaction with another PreBind plugin which can lead to this.

I've created https://github.com/kubernetes/kubernetes/issues/124024 for adding such an E2E test.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go [hunk 10]
```diff
@@ -701,7 +1125,24 @@ func (tc *testContext) verify(t *testing.T, expected result, initialObjects []me
 	sortObjects(wantObjects)
 	stripObjects(wantObjects)
 	stripObjects(objects)
-	assert.Equal(t, wantObjects, objects)
+	// Sometimes assert strips the diff too much, let's do it ourselves...
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1128]: No comment ;-}

File: pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+		Selector: "true",
+	}
+	filterNone := &resourceapi.NamedResourcesFilter{
+		Selector: "false",
+	}
+	filterBrokenType := &resourceapi.NamedResourcesFilter{
+		Selector: "1",
+	}
+	filterBrokenEvaluation := &resourceapi.NamedResourcesFilter{
+		Selector: `attributes.bool["no-such-attribute"]`,
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go:127]: Added. The CEL helper package itself has more complex tests for attributes.

File: pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+	}
+
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			tCtx := ktesting.Init(t)
+
+			controller, allocateErr := NewClaimController(tc.filter, tc.requests)
+			if allocateErr != nil {
+				if tc.expectCreateErr == "" {
+					tCtx.Fatalf("unexpected create error: %v", allocateErr)
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go:262]: `require.Error` would not have handled the "no error expected" case, so some additional if checks would have been needed and therefore I am writing out the checks manually here.

I tend to prefer `require + assert` whenever they are usable because that reduces the amount of test code and (IMHO) is more readable.

File: pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+				}
+				require.Equal(tCtx, tc.expectCreateErr, allocateErr.Error())
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go:264]: Isn't that a weaker check that what I have right now? I want to test the error explanation to ensure that it is readable and came from the right code path, not the error type.

This package doesn't have pre-defined errors, so `errors.Is` has nothing to compare against.

File: pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+
+			expectAllocateErr: "evaluate request CEL expression: no such key: no-such-attribute",
+		},
+	}
+
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			tCtx := ktesting.Init(t)
+
+			controller, allocateErr := NewClaimController(tc.filter, tc.requests)
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go:259]: Fixed.

File: pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go [hunk 15]
```diff
@@ -0,0 +1,1259 @@
...
+			// Not an error. It is safe to ignore unknown resources until a claim requests them.
+			// The unknown model in that claim then triggers an error for that claim.
+			expectResources: resources{"node": map[string]ResourceModels{
+				"driver": {
+					NamedResources: namedresourcesmodel.Model{},
+				},
+			}},
+		},
+
+		"one": {
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go:73]: Avoiding some variables is okay because their values are short, but repeating for example `twoInstances` whenever it's used IMHO would make it hard to see what the difference between two test cases is.

[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go:73]: I thought you were still referring to `namedresourcesmodel_test.go` because that is where I define common values in variables. Here I don't, but perhaps I should? It's not clear whether you call this out as a good example for tests or want me to change something :sweat_smile: 

"one" -> "one-instance"

File: pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go [hunk 16]
```diff
@@ -0,0 +1,1259 @@
...
+)
+
+func TestModel(t *testing.T) {
+	testcases := map[string]struct {
+		slices   resourceSliceLister
+		claims   assumeCacheLister
+		inFlight map[types.UID]resourceapi.ResourceClaimStatus
+
+		expectResources resources
+		expectErr       string
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go:45]: Changed.

File: pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go [hunk 17]
```diff
@@ -0,0 +1,1259 @@
...
+
+			var inFlightAllocations sync.Map
+			for uid, claimStatus := range tc.inFlight {
+				inFlightAllocations.Store(uid, &resourceapi.ResourceClaim{Status: claimStatus})
+			}
+
+			slices := tc.slices
+			if slices == nil {
+				slices = sliceList{}
+			}
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go:944]: `slices` is an interface because some tests use `sliceList` and others use `sliceError`. The interface value must not be nil, because then there is no implementation.

File: pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go [hunk 18]
```diff
@@ -0,0 +1,1259 @@
...
+			claims := tc.claims
+			if claims == nil {
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go:946]: See https://github.com/kubernetes/kubernetes/pull/123938#discussion_r1535252388

File: pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go [hunk 19]
```diff
@@ -0,0 +1,1259 @@
...
+				claims = claimList{}
+			}
+			actualResources, actualErr := newResourceModel(tCtx.Logger(), slices, claims, &inFlightAllocations)
+
+			if actualErr != nil {
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structuredparameters_test.go:951]: See https://github.com/kubernetes/kubernetes/pull/123938#discussion_r1535231175

File: test/e2e/dra/deploy.go (hunk 6 above)
Comment (pohly) [reply test/e2e/dra/deploy.go:95]: I want to do a full pass through this file and update it to use ktesting a bit more, but not now.

File: test/e2e/dra/dra.go (hunk 8 above)
Comment (pohly) [reply test/e2e/dra/dra.go:932]: I wanted names to be sortable in a sane way also when there are more than 10 instances, so I extended the naming with one leading 0. Probably doesn't matter much right now.

**Comment by pohly** [comment]
/kind bug

There is a (minor) fix included here, see first commit.

File: pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go]: > Isn't that a weaker check that what I have right now? 

It depends on the implementation of `Is`.

[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go]: Let's take an example. I want to check that the unknown attribute is included in the error message:
> `evaluate filter CEL expression: no such key: no-such-attribute`

This error is the result of `fmt.Errorf`. Are you suggesting that I should use a custom error type with an `error.Is` implementation which compares the error text? That seems a bit unusual to me. Nothing in the code depends on that behavior, it is only for testing.

[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go]: Or are you suggesting that the unit tests are too strict and shouldn't cover specific error texts? Then the test cases only specify a boolean for "expect some error".

Both approaches are used in various places. Some tests check that an error contains some substring. I'm leaning towards testing a bit more and checking that the full result of formatting and wrapping looks good, but I don't feel strongly about this.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go]: > Or are you suggesting that the unit tests are too strict and shouldn't cover specific error texts?

Kind of. We shouldn't need to update the unit tests if we change error messages a little. Is there more than one error possible? If not, maybe the "boolean" approach is ok.

Otherwise, we could define an error type (maybe called `CatchAll`) in some util package that implements `Is` returning true for any error. And once it's necessary to differentiate, the developer can add their own type or pointer to a specific error.

> Both approaches are used in various places. Some tests check that an error contains some substring

Yeah, I've trying to get rid of those, but there are just too many. In the past, comparing errors was harder because `Is` and `%w` didn't exist.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go]: Btw, the above intent is documented here https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md#technical-and-style-guidelines

[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go]: > Is there more than one error possible?

There is only a single error type. Only the explanation is different.

> If not, maybe the "boolean" approach is ok.

I've gone for that.

**Comment by alculquicondor** [comment]
We missed the test freeze, so this is no longer candidate for going to 1.30 or backporting.

But maybe we can backport the bugfix, if it's critical. How would you qualify it? If it's critical, split the fix into its own PR to go for 1.30.1

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Comment by pohly** [comment]
The bug is highly unlikely to occur in practice. Given that the feature is alpha and new in 1.30, I think we don't need to backport.
